File: Services/ParticipantService.cs
using System;
using System.Collections.Generic;
using TeamWeave.Models;

namespace TeamWeave.Services
{
    public class ParticipantService
    {
        private readonly List<Participant> _participants = new List<Participant>();
        private int _nextId = 1;

        /// <summary>
        /// Registers a new participant.
        /// </summary>
        public Participant RegisterParticipant(string name, string email, List<string> skillNames)
        {
            if (string.IsNullOrWhiteSpace(name) || string.IsNullOrWhiteSpace(email))
                throw new ArgumentException("Name and email are required");

            if (skillNames == null || skillNames.Count == 0)
                throw new ArgumentException("At least one skill is required");

            var participant = new Participant
            {
                Id = _nextId++,
                Name = name,
                Email = email,
                ProfileStatus = "Submitted"
            };

            _participants.Add(participant);
            return participant;
        }

        /// <summary>
        /// Adds a skill to a participant.
        /// </summary>
        public void AddSkillToParticipant(int participantId, Skill skill, int proficiencyLevel)
        {
            var participant = GetParticipantById(participantId);
            if (participant == null)
                throw new InvalidOperationException("Participant not found");

            if (proficiencyLevel < 1 || proficiencyLevel > 5)
                throw new ArgumentException("Proficiency level must be between 1 and 5");

            var participantSkill = new ParticipantSkill
            {
                ParticipantId = participantId,
                SkillId = skill.Id,
                ProficiencyLevel = proficiencyLevel,
                Skill = skill,
                Participant = participant
            };

            participant.Skills.Add(participantSkill);
            participant.UpdatedAt = DateTime.Now;
        }

        /// <summary>
        /// Adds interests to a participant.
        /// </summary>
        public void AddInterestsToParticipant(int participantId, List<string> interests)
        {
            var participant = GetParticipantById(participantId);
            if (participant == null)
                throw new InvalidOperationException("Participant not found");

            foreach (var interest in interests)
            {
                participant.Interests.Add(interest);
            }

            participant.UpdatedAt = DateTime.Now;
        }

        /// <summary>
        /// Retrieves a participant by ID.
        /// </summary>
        public Participant GetParticipantById(int id)
        {
            foreach (var participant in _participants)
            {
                if (participant.Id == id)
                    return participant;
            }
            return null;
        }

        /// <summary>
        /// Retrieves all participants.
        /// </summary>
        public List<Participant> GetAllParticipants()
        {
            return new List<Participant>(_participants);
        }

        /// <summary>
        /// Retrieves participants by skill category.
        /// </summary>
        public List<Participant> GetParticipantsBySkillCategory(string category)
        {
            var result = new List<Participant>();

            foreach (var participant in _participants)
            {
                foreach (var skill in participant.Skills)
                {
                    if (skill.Skill.Category == category)
                    {
                        result.Add(participant);
                        break;
                    }
                }
            }

            return result;
        }

        /// <summary>
        /// Updates participant profile status.
        /// </summary>
        public void UpdateParticipantStatus(int participantId, string newStatus)
        {
            var participant = GetParticipantById(participantId);
            if (participant == null)
                throw new InvalidOperationException("Participant not found");

            participant.ProfileStatus = newStatus;
            participant.UpdatedAt = DateTime.Now;
        }
    }
}

File: Services/ParticipantService.js
import { Participant } from '../Models/Participant.js';
import { AuditLog } from '../Models/AuditLog.js';

export class ParticipantService {
  /**
   * Registers a new participant in MongoDB.
   */
  static async registerParticipant({ name, email, skills = [], interests = [] }) {
    if (!name || !email) {
      throw new Error('Name and email are required');
    }
    if (!skills || skills.length === 0) {
      throw new Error('At least one skill is required');
    }

    // Check if participant already exists
    const existing = await Participant.findOne({ email: email.toLowerCase().trim() });
    if (existing) {
      throw new Error(`Participant with email ${email} already registered`);
    }

    const participant = new Participant({
      name: name.trim(),
      email: email.toLowerCase().trim(),
      skills: skills.map(s => ({
        skillName: s.name || s.skillName,
        category: s.category || 'Frontend',
        proficiencyLevel: Number(s.proficiency || s.proficiencyLevel || 3),
      })),
      interests: interests || [],
      profileStatus: 'Submitted',
    });

    const saved = await participant.save();

    // Audit log in MongoDB
    await AuditLog.create({
      changeType: 'ParticipantRegistered',
      participantId: saved._id,
      actor: 'Participant',
      details: `Participant '${saved.name}' registered with ${saved.skills.length} skills.`,
    });

    return saved;
  }

  /**
   * Retrieves all participants with optional skill category filter.
   */
  static async getAllParticipants(category) {
    if (category && category.trim() !== '') {
      return await Participant.find({ 'skills.category': category }).sort({ createdAt: -1 });
    }
    return await Participant.find().sort({ createdAt: -1 });
  }

  /**
   * Retrieves a single participant by MongoDB ObjectId.
   */
  static async getParticipantById(id) {
    return await Participant.findById(id);
  }

  /**
   * Updates participant profile status.
   */
  static async updateParticipantStatus(id, newStatus, actor = 'Organizer') {
    const participant = await Participant.findById(id);
    if (!participant) {
      throw new Error('Participant not found');
    }

    const oldStatus = participant.profileStatus;
    participant.profileStatus = newStatus;
    const updated = await participant.save();

    await AuditLog.create({
      changeType: 'ParticipantUpdated',
      participantId: participant._id,
      actor,
      oldValue: oldStatus,
      newValue: newStatus,
      details: `Updated status for ${participant.name}`,
    });

    return updated;
  }

  /**
   * Adds or updates a skill for a participant.
   */
  static async addSkillToParticipant(id, skillName, category, proficiencyLevel = 3) {
    const participant = await Participant.findById(id);
    if (!participant) {
      throw new Error('Participant not found');
    }

    const existingSkillIndex = participant.skills.findIndex(s => s.skillName.toLowerCase() === skillName.toLowerCase());
    if (existingSkillIndex > -1) {
      participant.skills[existingSkillIndex].proficiencyLevel = proficiencyLevel;
      participant.skills[existingSkillIndex].category = category;
    } else {
      participant.skills.push({
        skillName,
        category,
        proficiencyLevel,
      });
    }

    return await participant.save();
  }
}
